articles/android-application-widgets-tutorial-in-java/index.md

```diff
@@ -0,0 +1,144 @@
+</LinearLayout>
+```
+
+**Step 4: Widget Java class**
+Open the `NewAppWidget.java` file. This class extends with the `AppWidgetProvider` class. `AppWidgetProvider` receives and handles broadcast events relevant to the App Widget. `AppWidgetProvider` class overrides the following methods.
+
+- `onUpdate()`: This method updates App Widgets on intervals defined in the `AppWidgetProviderInfo Metadata`. In this case, the `AppWidgetProviderInfo Metadata` is defined in the file `new_app_widget_info.xml`. This method is also called when the App Widget is dropped on the home screen.
+- `onAppWidgetOptionsChanged()`: This method is called when the widget is resized.
+- `onEnabled()`: This method is called when a widget is created.
+`onDisabled()`: This method is called when the widget's last instance is deleted. For example, we can launch an activity within this function.
+
+**Step 5: Edit Manifest file**
+Declare the AppWidgetProvider class in the manifest file `AndroidManifest.xml` as a broadcast receiver. The IDE has already added this. If we created the widget class manually, we would have to add it manually. Refer to the `AndroidManifest.xml` in the [final project](https://github.com/manmusa100/App-Widgets-Demo).
+
+**Step 6: Edit widget configurations**
+In this step, we will change the widget cover image. Open the file `new_app_widget_info.xml`. The file contains widget configuration settings. Edit the line:
+
+```xml
+android:previewImage="@drawable/example_appwidget_preview"
+```
+Add a drawable image of your choice. Replace the `example_appwidget_preview` image with the image you added to the drawable folder. This image will act as the widget cover photo.
+
+**Step 7: Run the App**
+Install the app on your test device and add the widget to the home screen. Note widgets are arranged in alphabetical order. Go to the widget section and locate the widget `App Widgets Demo`. The widget we created, `NewAppWidget`, will have the same name as the application name, as shown in the screenshot below.
+
+![Demo widget screenshot](/engineering-education/android-application-widgets-tutorial-in-java/demo-widget-screenshot.jpg)
+
+**Step 8: Place the widget on the home screen**
+See the screenshot below of the final widget on a home screen.
+
+![Homescreen demo widget](/engineering-education/android-application-widgets-tutorial-in-java/home-screen-demo-widget.jpg)
+
+Tap the widget button. This should open the app `MainActivity`.
+
+### Limitations of Widgets
+Widgets have some limitations, including:
+- **Gestures:** Widgets support Touch and vertical swipe gestures. They don't support swiping laterally.
+- **Building blocks:** Due to the limited number of gestures, widgets cannot support some elements. Refer to the [Android widget documentation](https://developer.android.com/guide/topics/appwidgets/index.html#CreatingLayout) for a list of supported views.
+
+Source code [Github](https://github.com/manmusa100/App-Widgets-Demo)
+
+### Conclusion
```

Comment:
You can have a sample APK and even a gif demo of the widget to illustrate what a reader should achieve after using this tutorial.